lib/galaxy/schema/tools.py

```diff
@@ -0,0 +1,125 @@
+import json
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Union,
+)
+
+from pydantic import (
+    ConfigDict,
+    Field,
+    field_validator,
+)
+
+from galaxy.schema.fields import DecodedDatabaseIdField
+from galaxy.schema.jobs import (
+    EncodedJobDetails,
+    ShowFullJobResponse,
+)
+from galaxy.schema.schema import (
+    JobSummary,
+    Model,
+)
+
+
+class ExecuteToolPayload(Model):
+    tool_id: Optional[Any] = Field(
+        default=None,
+        title="Tool ID",
+        description="The ID of the tool to execute.",
+    )
+    tool_uuid: Optional[Any] = Field(
+        default=None,
+        title="Tool UUID",
+        description="The UUID of the tool to execute.",
+    )
+    action: Optional[str] = Field(
+        default=None,
+        title="Action",
+        description="The action to perform",
+    )
+    tool_version: Optional[str] = Field(
+        default=None,
+        title="Tool Version",
+        description="The version of the tool",
+    )
+    history_id: Optional[DecodedDatabaseIdField] = Field(
+        default=None,
+        title="History ID",
+        description="The ID of the history",
+    )
+
+    @field_validator("inputs", mode="before", check_fields=False)
+    @classmethod
+    def inputs_string_to_json(cls, v):
+        if isinstance(v, str):
+            return json.loads(v)
+        return v
+
+    inputs: Dict[str, Any] = Field(
+        default={},
+        title="Inputs",
+        description="The input data",
+    )
+    use_cached_job: Optional[bool] = Field(
+        default=False,
+        title="Use Cached Job",
+        description="Flag indicating whether to use a cached job",
+    )
+    preferred_object_store_id: Optional[str] = Field(
+        default=None,
+        title="Preferred Object Store ID",
+        description="The ID of the preferred object store",
+    )
+    input_format: Optional[str] = Field(
+        default="legacy",
+        title="Input Format",
+        description="The format of the input",
+    )
+    data_manager_mode: Optional[str] = Field(
+        default=None,
+        title="Data Manager Mode",
+        description="The mode of the data manager",
+    )
+    send_email_notification: Optional[bool] = Field(
+        default=None,
+        title="Send Email Notification",
+        description="Flag indicating whether to send email notification",
+    )
+    upload_type: Optional[str] = Field(
+        default=None,
+        title="Upload Type",
+        description="TODO",
+    )
+    model_config = ConfigDict(extra="allow")
```

Comment:
Do you expect any other field to be here that cannot be part of the model? If not we should avoid using `extra=allow`.

Reply:
Yes I expect other fields. They fields will have the prefix "files_" and it is not possible to narrow them down further (I think). The execute method in the service layer handles them like this:

```
    def execute(
        self,
        trans: ProvidesHistoryContext,
        payload: ExecuteToolPayload,
    ) -> ToolResponse:
        tool_id = payload.tool_id
        tool_uuid = payload.tool_uuid
        if tool_id in PROTECTED_TOOLS:
            raise HTTPException(
                status_code=400, detail=f"Cannot execute tool [{tool_id}] directly, must use alternative endpoint."
            )
        if tool_id is None and tool_uuid is None:
            raise HTTPException(status_code=400, detail="Must specify a valid tool_id to use this endpoint.")
        create_payload = payload.model_dump(exclude_unset=True)

        # process files, when they come in as multipart file data
        files = {}
        for key in list(create_payload.keys()):
            if key.startswith("files_") and isinstance(create_payload[key], UploadFile):
                files[key] = self.create_temp_file_execute(trans, create_payload.pop(key))
        create_payload.update(files)
        return self._create(trans, create_payload)
```